aetcd/client.py

```diff
@@ -154,47 +175,94 @@ def _init_channel_attrs(self):
+                    # Another task is establishing a connection. Just wait.
+                    await asyncio.wait_for(
+                        self._connected.wait(),
+                        self._connection_check_timeout,
+                    )
+                    return
+                break
+            except asyncio.TimeoutError:
+                # Another task fails to connect. Try again
+                continue
+        else:
+            raise asyncio.TimeoutError()
 
-        cred_params = [c is not None for c in (self._username, self._password)]
-        if all(cred_params):
-            self.auth_stub = rpc.AuthStub(self.channel)
-            auth_request = rpc.AuthenticateRequest(
-                name=self._username,
-                password=self._password,
-            )
+        try:
+            self._on_connecting = True
+            target = f'{self._host}:{self._port}'
+            self.channel = rpc.insecure_channel(target, options=self._options.items())
+
+            cred_params = [c is not None for c in (self._username, self._password)]
+            if all(cred_params):
+                self.auth_stub = rpc.AuthStub(self.channel)
+                auth_request = rpc.AuthenticateRequest(
+                    name=self._username,
+                    password=self._password,
+                )
 
-            resp = await self.auth_stub.Authenticate(auth_request, timeout=self._timeout)
-            self.metadata = (('token', resp.token),)
+                resp = await self.auth_stub.Authenticate(auth_request, timeout=self._timeout)
+                self.metadata = (('token', resp.token),)
 
-        self.kvstub = rpc.KVStub(self.channel)
-        self.clusterstub = rpc.ClusterStub(self.channel)
-        self.leasestub = rpc.LeaseStub(self.channel)
-        self.maintenancestub = rpc.MaintenanceStub(self.channel)
+            self.kvstub = rpc.KVStub(self.channel)
+            self.clusterstub = rpc.ClusterStub(self.channel)
+            self.leasestub = rpc.LeaseStub(self.channel)
+            self.maintenancestub = rpc.MaintenanceStub(self.channel)
 
-        # Initialize a watcher
-        self._watcher = watcher.Watcher(
-            rpc.WatchStub(self.channel),
-            timeout=self._timeout,
-        )
+            # Initialize a watcher
+            self._watcher = watcher.Watcher(
+                rpc.WatchStub(self.channel),
+                timeout=self._timeout,
+                metadata=self.metadata,
+            )
+            self._connected.set()
+        finally:
+            self._on_connecting = False
 
     async def close(self) -> None:
-        """Close established connection and frees allocated resources."""
+        """Close established connection and frees allocated resources.
+
+        NOTE: It could be called while other operation is being executed.
+              At that time, the executing operation will be broken.
+        """
+        for _ in range(MAX_COUNT_TO_WAIT_CONNECT_CHECK):
```

Comment:
It's long time ago for me to update it. Sorry I cannot remember the exact reason.
Probably `_connection_check_timeout` could be enough.

Reply:
Now I remember that it is for etcd cluster.
`MAX_COUNT_TO_WAIT_CONNECT_CHECK` is used for making a connection.
One host name is used for multiple etcd instances in a cluster. So I'd like to make multiple trial.
But here, it could not be required.

Reply:
I've created an issue  #20 to natively support multiple endpoints.